### PR TITLE
Support base controller method override

### PR DESCRIPTION
**Pull request by valeriob**
I would like to have a base area controller to aggregate common stuff : 
`   [AdminArea]
    public partial class AdminController : Controller
    {
        public override void OnActionExecuting(ActionExecutingContext context)
        {
            base.OnActionExecuting(context);
        }
    }`

I excluded members returning void, is it good enough ?

### CODE REVIEW

**Comment by artiomchi** [comment]
Good start for sure. Is the purpose of this to just host the Area attribute, or more to have common base helpers (which means this will also affect non area controllers)?

Regarding the `AdminController` you have there. If that's going to be a base controller, I would usually call it `AdminControllerBase` or something similar. But wouldn't that class be abstract? You don't want the MVC framework to actually pick it up as a separate controller, as well as have all the controllers that inherited it to also have those same methods.

Regarding the methods themselves - R4Mvc should see any method that is going to be picked up by the MVC framework itself.. So skipping the `[NonAction]` methods makes perfect sense, since the framework will skip them too. The other methods, on the other hand?

- `NonActionMethodVoidReturn` - it's already skipped since it's protected. It won't be picked by either the framework or R4Mvc
- `NonActionMethodWithReturnValue` - same as above
- `PublicNonActionMethodVoidReturn` that's a perfectly valid method. If requested, MVC will run it, and return a 0 length 200 OK response. As such, I see no reason why it should be skipped :)
- `PublicNonActionMethodMustBeExcludedWithNonAction` - `[NonAction]` decorated method - perfect reason to skip the method :)

---

Long story short, I agree with the [NonAction] filtering (although I'd name the method something like `IsNotNonActionExcluded`) and Base type AreaAttribute lookup. But I don't think we should skip void return methods.

**Comment by valeriob** [comment]
Thanks @artiomchi.
The purpose is to write the code i've shown, so yes expose some common functionality via base controller for the area, and customize some area stuff on the layout etc.

Yes it should be abstract, but testing r4mvc i saw no differences, i think it still generates stuff for it, i'll check it out.

I've added those methods as "documentation" i wanted to write them on the test, but i don't know how to test this thing yet.

The reason i skipped the return void methods is because they are never picked up by mvc, and r4mvc kept producing overrides for   public override void OnActionExecuting(ActionExecutingContext context).
How do you prefer to address this ?

**Comment by artiomchi** [comment]
Hmm, I'll have to check in the evening, but if R4Mvc is processing abstract classes then that's a bug :)

> I've added those methods as "documentation" i wanted to write them on the test, but i don't know how to test this thing yet.

Oh yeah, the methods are great in there.. I'm working on refactoring at the moment before I start adding tests (I should actually transfer my notes to an issue here from my notepad), but I was doing the same thing - I'm using the host project as a testbed to test the functionality as I'm developing, and the unit tests will be used to prevent existing functionality from being broken by new code changes :)

> The reason i skipped the return void methods is because they are never picked up by mvc

Oh yes they are! I just tested it myself! That's why I don't think we should skip them

> r4mvc kept producing overrides for public override void OnActionExecuting(ActionExecutingContext context)

Hmm... Now that's an interesting one.. We might have to add a list of `Controller` class methods that should be ignored by R4Mvc. Possibly by getting a list of methods from the `Controller` class using reflection.

In T4MVC they just look at methods returning `ActionResults`. In R4Mvc, we should look at any return types, since AspNetCore mixes MVC with WebApi in one project. So that's my best idea on how to handle it so far.

Thoughts?

**Comment by valeriob** [comment]
> Oh yes they are! I just tested it myself! That's why I don't think we should skip them
Oh, i never wrote that kind of code.

> In T4MVC they just look at methods returning ActionResults. In R4Mvc, we should look at any return types, since AspNetCore mixes MVC with WebApi in one project. So that's my best idea on how to handle it so far.

I tried to look for the assembly where a method is defined and exclude those on Microsoft.AspNetCore.MVC but i could not do it, maybe i should have insisted more.
Another solution would be to black list method names, but i would not do it via reflection, just statically compile a list with strings, and maybe add the possibility for the users to manipulate it for future compatibility without requiring a new release or something.

**Comment by artiomchi** [comment]
> statically compile a list with strings, and maybe add the possibility for the users to manipulate it for future compatibility without requiring a new release or something.

Hmm... Statically compile the list? Sure, maybe.. But that would mean if MVC adds another public method to the `Controller` class, we'd catch it, and we'd have to make a new release to counter that.

Let users manipulate the list? Not a good idea for two reasons:

- Say a new method is added and someone blacklists that in their code. It'll only affect that one user, everyone else won't benefit from that.
- MVC releases don't happen too often. Releases that would add a public method to the Controller class? Almost never happens. Adding a customisation feature for this would have almost no use :)

Besides, the reflection part would only be used in the tools project, i.e. during R4Mvc generation. Never at runtime, so it'll never affect the end project performance. I still think it's not a bad solution, and it'll be self serving - any new methods added by MS would be automatically picked up and filtered straight away during every R4Mvc generation

**Comment by valeriob** [comment]
> Besides, the reflection part would only be used in the tools project, i.e. during R4Mvc generation. Never at runtime, so it'll never affect the end project performance.

Good point, i just wanted to avoid runtime reflection 😄

**Comment by artiomchi** [comment]
Ooooh yeah, runtime reflection here? Definitely not a good idea.

That's one of the reason I love T4MVC/R4Mvc - they don't change the framework, and they don't carry almost any overhead. And with the raw perf that AspNetCore brings, I would not want this project to just ruin all that :)

**Review by artiomchi — CHANGES_REQUESTED** [review]
So yeah, let's get the `[NonAction]` and base class area lookup in as a minimum. And you can add the `Controller` class public method filtering as well.

How does that sound?

**Comment by artiomchi** [comment]
By the way, you were right about R4Mvc processing abstract controllers - I just patched it as promised :)

**Comment by valeriob** [comment]
Thanks for the fix !
So with this last commit we avoid to rewrite and generate helper methods for any method inherited from Microsoft.AspNetCore.Mvc.Controller or marked as NonAction

**Review by artiomchi — CHANGES_REQUESTED** [review]
Moving in the right direction, yes, but you've WAY overcomplicated it :)

A couple tips:
- The `CompilationSettings` class and passing around the `mvcMethodNames` is completely unnecessary. That means you have to start passing that as a parameter to everything, plus it's not supposed to be configurable anyway. Ultimately, it's only accessed directly in a single place (`IsAction` method, which I'd rename to `IsMvcAction` btw), so I'd rather it got generated there on first access
- The actual Controller method population is not bad (although could definitely be simplified and optimised), but it should only look at virtual methods, no?
- In `ControllerRewriter.cs`, the check that ignores abstract methods is redundant, since you can't have abstract methods in non abstract classes
- As far as the host project is concerned, in the `UsersController`, the `Edit` and `PublicNonActionMethodMustBeExcludedWithNonAction` methods don't really add anything, since `ActionExcluded` already handles this change.
- Same controller, I think it would be nicer to use a different base controller method as a test, say `OnActionExecuted` to showcase that base controller methods will be ignored whether they come from the base class, or inside the controller itself.
- Also, I recommend you try keep your code formatting cleaner, as you end up adding a whole bunch of line breaks in random places :)

Sorry if I'm a stickler to details, but I'm trying to get the code as lean as I can (hence the current alpha 1.1 work is just code refactoring, cleaning and optimisation, now that we have a working prototype). 

Ultimately, this PR should only have minor changes in 3 files in the tools project, everything else can be reverted. Ultimately, I could do it myself, but I don't want to prevent anyone from contributing and hogging all the commits to myself :D

**Comment by valeriob** [comment]
No problem at all!

I passed the list around because you started to use DI for services, there are some interface to get around and I saw that R4MvcGenerator is not a singleton, so mvcMethodNames is a function of it's parameter CSharpCompilation, you can invoke R4MvcGenerator  on two different project with two different aspnetcore versions.

I guess it's going to take far more time to me to do it than you, sorry for making you waste time in the review, do it as you prefer as far as the features are preserved, i'm glad i moved the conversation forward 😄

**Comment by artiomchi** [comment]
No worries about the time spent, we did add the features (and found some bugs in the process!)

I also found another bug that actually plagues Alpha builds right now, so that's great too!
